sources/api/shibaken/src/warmpool/autoscaling_warm_pool.rs

```diff
@@ -0,0 +1,146 @@
+/// This module contains utilities for waiting until the instance on warm pool to be
+/// marked as InService.
+use crate::warmpool::error::{self, Result};
+use argh::FromArgs;
+use imdsclient::ImdsClient;
+use serde::Deserialize;
+use snafu::{OptionExt, ResultExt};
+use std::fs;
+use std::path::Path;
+use tokio::time::{sleep, Duration};
+
+//Retry strategy variables
+const FETCH_IMDS_SLEEP_DURATION: Duration = Duration::from_secs(5);
+const MAX_ATTEMPTS: i8 = 5;
+
+// Path to config file containing host's autoscaling.should-wait setting
+const CONFIG_PATH: &str = "/etc/warm-pool-wait.toml";
+const FETCH_LIFECYCLE_INTERVAL_SECS: Duration = Duration::from_secs(5);
+
+#[derive(FromArgs, Debug)]
+#[argh(subcommand, name = "warm-pool-wait")]
+/// Wait until the instance reaches the `InService` state
+pub(crate) struct WarmPoolWait {}
+
+impl WarmPoolWait {
+    pub(crate) async fn run(self) -> Result<()> {
+        let config_parse = get_config().await?;
+
+        let should_wait_value = config_parse.should_wait;
+        let marker_file_path = config_parse.marker_path;
+
+        println!("autoscaling.should-wait value is {}", should_wait_value);
+        if should_wait_value {
+            wait_until_inservice().await?;
+        }
+
+        fs::write(&marker_file_path, "").unwrap_or_else(|e| {
+            log::warn!("Failed to create marker file '{}', warm-pool-wait service may unexpectedly run again: '{}'",
+            &marker_file_path, e);
+        });
+        println!("Marker file path is {}", marker_file_path);
+
+        Ok(())
+    }
+}
+
+/// Uses an imdsclient function to fetch the lifecyclestate of the host
+async fn get_lifecycle_state(client: &mut ImdsClient) -> Result<Option<String>> {
+    client
+        .fetch_autoscaling_lifecycle_state()
+        .await
+        .context(error::ImdsRequestSnafu)
+}
+
+/// Continuously fetches lifecycle state of the host.
+/// Returns from function when host is "InService"
+async fn wait_until_inservice() -> Result<()> {
```

Comment:
This still seems more complicated than what you need. If we're willing to wait forever for the state to reach "InService", it seems like we should also be willing to wait forever for IMDS not to return 404.

Accordingly, what I had in mind was:
```
/// Continuously fetches lifecycle state of the host.
/// Returns from function when host is "InService"
async fn wait_until_inservice() -> Result<()> {
    let mut client = ImdsClient::new();

    loop {
        let lifecycle_state = client
            .fetch_autoscaling_lifecycle_state()
            .await
            .context(error::ImdsRequestSnafu)?;
        if let Some(lifecycle_state) = lifecycle_state {
            if lifecycle_state == "InService" {
                println!("Lifecycle state is {} ... exiting.", lifecycle_state);
                return Ok(());
            }
        }
        sleep(FETCH_LIFECYCLE_INTERVAL_SECS).await;
        continue;
    }
}
```

Here's the [git diff](https://gist.github.com/bcressey/2964cfaa2af04bbc8f5b02fb93570de0) output from my edits if that's useful.